services/src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -43,71 +43,89 @@
     _topic = config.getString(ProduceServiceConfig.TOPIC_CONFIG);
     _producerId = config.getString(ProduceServiceConfig.PRODUCER_ID_CONFIG);
     _produceDelayMs = config.getInt(ProduceServiceConfig.PRODUCE_RECORD_DELAY_MS_CONFIG);
     _recordSize = config.getInt(ProduceServiceConfig.PRODUCE_RECORD_SIZE_BYTE_CONFIG);
     _sync = config.getBoolean(ProduceServiceConfig.PRODUCE_SYNC_CONFIG);
     _partitionNum = new AtomicInteger(0);
+    _running = new AtomicBoolean(false);
+    _nextIndexPerPartition = new ConcurrentHashMap<>();
 
-    if (_rebalanceThreshold < 1) {
-      throw new IllegalArgumentException("Rebalance threshold must be greater than one but is set to " + _rebalanceThreshold + ".");
-    }
-
+    int autoTopicPartitionFactor = config.getInt(ProduceServiceConfig.PARTITIONS_PER_BROKER_CONFIG);
+    int autoTopicReplicationFactor = config.getInt(ProduceServiceConfig.TOPIC_REPLICATION_FACTOR_CONFIG);
     int existingPartitionCount = Utils.getPartitionNumForTopic(_zkConnect, _topic);
 
     if (existingPartitionCount <= 0) {
-      if (config.getBoolean(ProduceServiceConfig.AUTO_TOPIC_CREATION_ENABLED_CONFIG)) {
-        int autoTopicReplicationFactor = config.getInt(ProduceServiceConfig.AUTO_TOPIC_REPLICATION_FACTOR_CONFIG);
-        int autoTopicPartitionFactor = config.getInt(ProduceServiceConfig.REBALANCE_PARTITION_MULTIPLE_CONFIG);
+      if (config.getBoolean(ProduceServiceConfig.TOPIC_CREATION_ENABLED_CONFIG)) {
         _partitionNum.set(
             Utils.createMonitoringTopicIfNotExists(_zkConnect, _topic, autoTopicReplicationFactor,
                 autoTopicPartitionFactor));
       } else {
         throw new RuntimeException("Can not find valid partition number for topic " + _topic +
             ". Please verify that the topic \"" + _topic + "\" has been created. Ideally the partition number should be" +
             " a multiple of number" +
-            " of brokers in the cluster.  Or else configure " + ProduceServiceConfig.AUTO_TOPIC_CREATION_ENABLED_CONFIG +
+            " of brokers in the cluster.  Or else configure " + ProduceServiceConfig.TOPIC_CREATION_ENABLED_CONFIG +
             " to be true.");
       }
     } else {
       _partitionNum.set(existingPartitionCount);
     }
 
+    initializeProducer(_producerPropsOverride, _producerClass);
+
+    _executor = Executors.newScheduledThreadPool(_threadsNum);
+    _checkAddPartitionsExecutor = Executors.newSingleThreadScheduledExecutor();
+
+    MetricConfig metricConfig = new MetricConfig().samples(60).timeWindow(1000, TimeUnit.MILLISECONDS);
+    List<MetricsReporter> reporters = new ArrayList<>();
+    reporters.add(new JmxReporter(JMX_PREFIX));
+    Metrics metrics = new Metrics(metricConfig, reporters, new SystemTime());
+    Map<String, String> tags = new HashMap<>();
+    tags.put("name", _name);
+    _sensors = new ProduceMetrics(metrics, tags);
+
+    _topicRebalanceService = new TopicManagementService(props, _name + "-topic-management-service");
```

Comment:
I've updated the name.    I thought the purpose of starting it inside was so that the configuration would be in the same block in the config file.  Otherwise configuration needs to be duplicated in different blocks.  Is this not correct?